src/smt.rs

```diff
@@ -1,163 +1,632 @@
+    }
+}
+
+pub fn bvsdiv(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvSDiv,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvurem(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvURem,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvsmod(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvSMod,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvult(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvULt,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvugt(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvUGt,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvslt(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvSLt,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvsgt(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvSGt,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvshl(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvShl,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvlshr(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvLShr,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvashr(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvAShr,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn extract(msb: u64, lsb: u64, arg: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Extract(msb, lsb),
+        args: vec![arg],
+    }
+}
+
+pub fn concat(args: Vec<SMTExpr>) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Concat,
+        args,
+    }
+}
+
+pub fn select(arr: SMTExpr, idx: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Select,
+        args: vec![arr, idx],
+    }
+}
+
+pub fn store(arr: SMTExpr, idx: SMTExpr, val: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Store,
+        args: vec![arr, idx, val],
+    }
+}
+
+pub fn literal(lit: String) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Literal(lit),
+        args: vec![],
+    }
+}
+
+pub fn uf(function: SMTVariable, args: Vec<SMTExpr>) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::UF(function),
+        args,
+    }
+}
+
+// SMT statement builders
+
+pub fn assert(expr: SMTExpr) -> SMTStatement {
+    SMTStatement::Assert(expr)
+}
+
+pub fn declare_const(var: SMTVariable) -> SMTStatement {
+    SMTStatement::DeclareConst(var)
+}
+
+pub fn declare_fun(var: SMTVariable, sorts: Vec<SMTSort>) -> SMTStatement {
+    SMTStatement::DeclareFun(var, sorts)
+}
+
+pub fn define_const(var: SMTVariable, val: SMTExpr) -> SMTStatement {
+    SMTStatement::DefineConst(var, val)
+}
+
+pub fn define_fun(var: SMTVariable, sorts: Vec<SMTSort>, val: SMTExpr) -> SMTStatement {
+    SMTStatement::DefineFun(var, sorts, val)
+}
+
+// Format stuff
+
+pub trait SMTFormat {
+    fn as_smt(&self) -> String;
+}
+
+impl SMTFormat for SMTSort {
```

Comment:
Didn't you want to replace this by a regular Display impl or something?

Reply:
Kept it for now cus was convenient, but also fine with removing